src/components/map/MapboxGlMap.jsx

```diff
@@ -112,7 +112,23 @@ export default class MapboxGlMap extends React.Component {
     const nav = new MapboxGl.NavigationControl();
     map.addControl(nav, 'top-right');
 
-    const inspect = new MapboxInspect({
+    map.on("style.load", () => {
+      const inspect = this.generateInspectControl()
+      map.addControl(inspect)
```

Comment:
@gregorywolanski some questions

 - can you explain why this works, I'm a little confused why this fixes the issue
 - do we also want to be running `map.removeControl` at some point.

Reply:
### Why this works?

English is not my primary language so don't hesitate to ask clarifying questions. :)

Bear with me: mapbox-gl-inspect (MapboxInspect) uses map styles. Map styles contain sources ("openstreetmap", "mapbox"). Right now, without the bug fix, Maputnik starts with one map style and invoke MapboxInspect only once. It works OK as long as you use map styles that use the same source. In this scenario, when you change map style, the source list remains the same. However, when you switch to a map style with a different source (as in the example provided by @muesliq), MapboxInspect is not updated, uses the old list of sources, which lacks the new source and the app returns `Error: layers: source "mapbox" not found`. The bug fix enables MapboxInspect to be updated every time the user change the style.

* * *

### Do we also want to be running `map.removeControl` at some point?

Good question. I haven't thought about that, because of no organoleptic signs that it can lead to problems.  I'll add today `map.removeControl` before `map.addControl` if `this.inspect` has been defined.

Reply:
> Right now, without the bug fix, Maputnik starts with one map style and invoke MapboxInspect only once. It works OK as long as you use map styles that use the same source.

Thanks for the context. So I think adding / removing sources from the current style also causes the same issue. You can see this if you follow the steps

 1. Load the empty map style
 2. Add openmaptiles source
 3. Go into inspect mode
 4. Remove openmaptiles source
 5. Go into inspect mode again

You'll see the source that no longer exists is still being inspected.

I think the problem actually lies in [lukasmartinelli/mapbox-gl-inspect](https://github.com/lukasmartinelli/mapbox-gl-inspect) because this does the tracking of changes to the sources list <https://github.com/lukasmartinelli/mapbox-gl-inspect/blob/master/lib/MapboxInspect.js#L203-L204>.

I think the real issue might be here <https://github.com/lukasmartinelli/mapbox-gl-inspect/blob/master/lib/MapboxInspect.js#L128-L148> but I'm unsure

Thoughts?

Reply:
Yes, you’re right, adding and removing sources from the current style also causes the same issue.

First observation: https://github.com/lukasmartinelli/mapbox-gl-inspect/blob/master/lib/MapboxInspect.js#L128-L148 looks like it’s about a different type of sources list. I mean: The subject of the bug is an array that looks like `["mapbox", "openstreetmap"]`. The piece of code that you mentioned operates on objects like `{openmaptiles: Array(15)}`. 

I’m digging into [lukasmartinelli/mapbox-gl-inspect](https://github.com/lukasmartinelli/mapbox-gl-inspect).

Reply:
> looks like it’s about a different type of sources list.

Yeah sorry having a quick look and pointing at random code probably wasn't that helpful :)

> I’m digging into lukasmartinelli/mapbox-gl-inspect.

 Yay! Just shout if you want me to test anything

Reply:
You were right, it’s a problem that can be solved inside [lukasmartinelli/mapbox-gl-inspect](https://github.com/lukasmartinelli/mapbox-gl-inspect).

It took me an embarrassing number of hours to debug, but I think that I found the problem and the solution.

I’ll get some sleep, test it once again, prepare a pull request in [lukasmartinelli/mapbox-gl-inspect](https://github.com/lukasmartinelli/mapbox-gl-inspect) and then close this pull request.

It’s a pleasure to work with you, @orangemug. :)

Reply:
> ...but I think that I found the problem and the solution

Yay, congrats @gregorywolanski! If you'd like me to give it a quick test just mention me in the new PR

Reply:
It turned out my solution doesn’t remove the empty map style problem. :/ :)

I’ll ask you for a quick test in a few days, thank you.